packages/core/src/hydration/event_replay.ts

```diff
@@ -79,3 +118,78 @@ export function setJSActionAttribute(
     }
   }
 }
+
+function setEventReplayer(dispatcher: Dispatcher) {
+  dispatcher.setEventReplayer(queue => {
+    for (const event of queue) {
+      handleEvent(event);
+    }
+  });
+}
+
+/**
+ * TODO: Update documentation here.
+ *
+ * @param target Component, Directive or DOM Node.
+ */
+export function getLViewByElement(target: any): LView|null {
+  let lView = readLViewData(target);
+  if (lView) {
+    return lView;
+  } else {
+    // if the context is not found then we need to traverse upwards up the DOM
+    // to find the nearest element that has already been monkey patched with data
+    let parent = target as any;
+    while (parent = parent.parentNode) {
+      lView = readLViewData(parent);
+
+      // the edge of the app was also reached here through another means
+      // (maybe because the DOM was changed manually).
+      if (lView) {
+        attachLViewId(target, lView);
+        return lView;
+      }
+    }
+  }
+  return null;
+}
+
+function handleEvent(event: EventInfoWrapper) {
+  const eventName = event.getEventType();
+  const nativeElement = event.getAction()!.element;
+  // Dispatch event via Angular's logic
+  if (nativeElement) {
+    const lView = getLViewByElement(nativeElement);
+    if (lView !== null) {
+      const tView = lView[TVIEW];
+      const listeners = getEventListeners(tView, lView, nativeElement, eventName);
+      for (const listener of listeners) {
+        listener(event.getEvent());
+      }
+    }
+  }
+}
```

Comment:
nit: we can move the `event.getEventType()` call into the branch where we need it (avoid unnecessary calls). I've also extracted the `event.getEvent()` outside of the loop (for same reasons - avoid extra calls).
```suggestion
  const nativeElement = event.getAction()!.element;
  // Dispatch event via Angular's logic
  if (nativeElement) {
    const lView = getLViewByElement(nativeElement);
    if (lView !== null) {
      const tView = lView[TVIEW];
      const eventName = event.getEventType();
      const origEvent = event.getEvent();
      const listeners = getEventListeners(tView, lView, nativeElement, eventName);
      for (const listener of listeners) {
        listener(origEvent);
      }
    }
  }
}
```

Reply:
done